### PR TITLE
Add review queues for unlisted addons (bug 1121237)

### PR DESCRIPTION
**Pull request by magopian**
Fixes [bug 1121237](https://bugzilla.mozilla.org/show_bug.cgi?id=1121237)

### CODE REVIEW

**Comment by magopian** [comment]
That's a huge one... @clouserw @kmaglione r?

File: apps/amo/fixtures/base/users.json [hunk 1]
```diff
@@ -191,6 +214,16 @@
         }
     },
     {
+        "pk": 50004,
+        "model": "access.group",
+        "fields": {
+            "rules": "Addons:Review,Addons:ReviewUnlisted,Reviews:Edit,ReviewerAdminTools:View",
+            "modified": "2012-01-27 13:32:08",
+            "name": "Senior Add-on Reviewers",
+            "created": "2012-01-27 12:41:00"
```
Comment (kmaglione) [inline apps/amo/fixtures/base/users.json]: +1 I've wanted this so many times.

File: apps/editors/decorators.py [hunk 2]
```diff
@@ -33,6 +33,23 @@ def wrapper(request, *args, **kw):
     return wrapper
 
 
+def unlisted_addons_reviewer_required(f):
+    """Requires the user to be logged in as an unlisted addons reviewer or
+    admin.
+
+    An unlisted addons reviewer is someone who is in a group with the following
+    permission: Addons:ReviewUnlisted.
+
```
Comment (kmaglione) [inline apps/editors/decorators.py]: Extra blank link, and missing newline after opening quotes. I know the other decorators are like this as well, but it makes me twitch.

File: apps/editors/helpers.py [hunk 3]
```diff
@@ -116,9 +121,18 @@ def editors_breadcrumbs(context, queue=None, addon_queue=None, items=None,
             'flagged_themes': _('Flagged Themes'),
             'rereview_themes': _('Update Themes'),
         }
+        if unlisted:
+            queues = {
```
Comment (kmaglione) [inline apps/editors/helpers.py]: I think I'd prefer an if-else here. It's not immediately obvious what's happening on first read.

File: apps/editors/helpers.py [hunk 4]
```diff
@@ -160,6 +175,22 @@ def queue_tabnav(context):
                (ngettext('Moderated Review ({0})', 'Moderated Reviews ({0})',
                          counts['moderated'])
                 .format(counts['moderated'])))]
+    if context.get('unlisted'):
```
Comment (kmaglione) [inline apps/editors/helpers.py]: I'd definitely prefer an if-else here, especially since it would save a needless queue counts query.

File: apps/editors/templates/editors/base.html [hunk 5]
```diff
@@ -38,24 +38,40 @@ <h1 class="site-title amo">
...
-                  {{ _('Fast Track') }} ({{ queue_counts['fast_track'] }})</a></li>
-            <li><a href="{{ url('editors.queue_nominated') }}">
-                  {{ _('Full Reviews') }} ({{ queue_counts['nominated'] }})</a></li>
-            <li><a href="{{ url('editors.queue_pending') }}">
-                  {{ _('Pending Updates') }} ({{ queue_counts['pending'] }})</a></li>
-            <li><a href="{{ url('editors.queue_prelim') }}">
-                  {{ _('Preliminary Reviews') }} ({{ queue_counts['prelim'] }})</a></li>
-            <li><a href="{{ url('editors.queue_moderated') }}">
-                  {{ _('Moderated Reviews') }} ({{ queue_counts['moderated'] }})</a></li>
+          {% if unlisted_queue_counts %}
```
Comment (kmaglione) [inline apps/editors/templates/editors/base.html]: Shouldn't this be outside of the `<ul>`?

Comment (kmaglione) [inline apps/editors/templates/editors/base.html]: (Same question for above, obviously)

File: apps/editors/templates/editors/queue.html [hunk 6]
```diff
@@ -27,7 +27,7 @@
         <button type="submit">{{ _('Search') }}</button>
         <a class="toggle-queue-search" href="#">{{ _('Advanced Search') }}</a>
         {% if search_form.data and search_form.data['searching'] %}
-          (<a class="clear-queue-search" href="{{ url('editors.queue_%s' % tab) }}">{{ _('clear search') }}</a>)
+        (<a class="clear-queue-search" href="{{ url('editors.unlisted_queue_{0}'.format(tab)) if unlisted else url('editors.queue_{0}'.format(tab)) }}">{{ _('clear search') }}</a>)
```
Comment (kmaglione) [inline apps/editors/templates/editors/queue.html]: This is a bit hard to parse. I think `{{ url('editors.{1}queue_{0}'.format(tab, ('unlisted_' if unlisted else ''))) }}` is slightly more readable, but not a big deal either way.

File: apps/editors/tests/test_views.py [hunk 7]
```diff
@@ -438,13 +456,79 @@ def test_new_editors(self):
...
+        # Both listed and unlisted queues for senior editors.
+        self.login_as_senior_editor()
+        doc = pq(self.client.get(self.url).content)
+        queues = doc('#listed-queues ul li a')
+        queues_links = [link.attrib['href'] for link in queues]
+        assert queues_links == listed_queues_links
+        unlisted_queues = doc('#unlisted-queues ul li a')
+        unlisted_queues_links = [link.attrib['href']
+                                 for link in unlisted_queues]
+        assert unlisted_queues_links == [
```
Comment (kmaglione) [inline apps/editors/tests/test_views.py]: It's a bit weird that `listed_queues_links` is a list of static URLs while `unlisted_queues_links` is a list of URLs from the page. Can you make them both lists of static URLs at the top of the functions so the two tests follow the same pattern?

File: apps/editors/tests/test_views.py [hunk 8]
```diff
@@ -660,23 +744,10 @@ def test_legacy_queue_sort(self):
...
-
-            doc = pq(r.content)
-            div = doc('#editors-stats-charts .editor-stats-table:eq(0)')
-
-            eq_(div('.waiting_old').attr('style'), style(widths[0]))
-            eq_(div('.waiting_med').attr('style'), style(widths[1]))
-            eq_(div('.waiting_new').attr('style'), style(widths[2]))
+        addon = self.addons['Nominated Two']
+        for data in self.get_review_data():
+            self.check_bar(addon, eq=0, data=data, reset_status=False)
```
Comment (kmaglione) [inline apps/editors/tests/test_views.py]: Nice.

File: apps/editors/views.py [hunk 9]
```diff
@@ -634,6 +678,9 @@ def reviewlog(request):
     form = forms.ReviewLogForm(data)
 
     approvals = ActivityLog.objects.review_queue()
+    if not acl.check_unlisted_addons_reviewer(request):
+        # Only display logs related to unlisted addons to senior reviewers.
```
Comment (kmaglione) [inline apps/editors/views.py]: Nit: This is a bit ambiguous. I think it means "Display logs related to unlisted add-ons only to senior reviewers"

**Comment by kmaglione** [comment]
Comments aside, looks good to me.

File: apps/editors/decorators.py (hunk 2 above)
Comment (magopian) [inline apps/editors/decorators.py]: There shouldn't be a newline after opening quotes for a docstring (single or multi-line), according to the [pep](https://www.python.org/dev/peps/pep-0257/#id17). However, the first line should also fit on a single line... which isn't the case here.

Also the wording isn't correct: taken from the pep
```The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...". ```

I'll change those.

File: apps/editors/templates/editors/base.html (hunk 5 above)
Comment (magopian) [inline apps/editors/templates/editors/base.html]: I've removed those... they were needed because the "abuse reports" page wasn't providing the proper context (with the queue counts). I've changed that, and thus removed those tests.

File: apps/editors/templates/editors/queue.html (hunk 6 above)
Comment (magopian) [inline apps/editors/templates/editors/queue.html]: Good idea, thanks

File: apps/editors/tests/test_views.py (hunk 7 above)
Comment (magopian) [inline apps/editors/tests/test_views.py]: Sure, that would make more sense, thanks

File: apps/editors/tests/test_views.py (hunk 8 above)
Comment (magopian) [inline apps/editors/tests/test_views.py]: Just a little bit of cleanup ;)

File: apps/editors/views.py (hunk 9 above)
Comment (magopian) [inline apps/editors/views.py]: Thanks ;)

**Comment by magopian** [comment]
Thanks a lot for the review!
